### DISCUSSION TITLE
HelmRelease stuck when helm-controller pod dies

### DISCUSSION THREAD

**Discussion by hscheib** [fluxcd/flux2#3555] [Q&A]
Hi,

We have a few HelmRelease that can take well over 10 minutes to finish upgrading. We have seen multiple times the HR gets perpetually stuck in a `reconciliation in progress` state even though all the pods finish rolling out successfully. 

We notice that during the deploy the helm-controller would get sent SIGTERM

```
Feb 2 11:02:15 helm-controller-68686dc594-rmfbm manager info {"level":"info","ts":"2023-02-02T17:02:15.683Z","msg":"Stopping and waiting for non leader election runnables"}
Feb 2 11:02:15 helm-controller-68686dc594-rmfbm manager info {"level":"info","ts":"2023-02-02T17:02:15.683Z","msg":"Stopping and waiting for leader election runnables"}
Feb 2 11:02:15 helm-controller-68686dc594-rmfbm manager info {"level":"info","ts":"2023-02-02T17:02:15.683Z","logger":"controller.helmrelease","msg":"Shutdown signal received, waiting for all workers to finish","reconciler group":"helm.toolkit.fluxcd.io","reconciler kind":"HelmRelease"}
Feb 2 11:02:45 helm-controller-68686dc594-rmfbm manager info {"level":"info","ts":"2023-02-02T17:02:45.684Z","msg":"Stopping and waiting for caches"}
Feb 2 11:02:45 helm-controller-68686dc594-rmfbm manager info {"level":"info","ts":"2023-02-02T17:02:45.684Z","msg":"Stopping and waiting for webhooks"}
Feb 2 11:02:45 helm-controller-68686dc594-rmfbm manager info {"level":"info","ts":"2023-02-02T17:02:45.684Z","msg":"Wait completed, proceeding to shutdown the manager"}
Feb 2 11:02:45 helm-controller-68686dc594-rmfbm manager error {"level":"error","ts":"2023-02-02T17:02:45.696Z","logger":"setup","msg":"problem running manager","error":"failed waiting for all runnables to end within grace period of 30s: context deadline exceeded"}
```

This is usually because the underlying node is being terminated and the pod needs to move to a new node. Since the HR mentioned takes so long to finish deploying, the new helm-controller pod that spins up doesn't know how to continue and the helm status is never updated to success.

We have a policy where nodes can only be around for N amount of time before we recycle them, so its something we will continue to see when the helm-controller is moved during a long running helm upgrade. It is in frequent but it has caused manual intervention multiple times now I am looking to help resolve via automation.

Is there a recommended strategy for handling this scenario? Can I increase the grace period for the helm-controller?

**Comment by karrth** [comment]
I believe this is fixed by these 2 MRs allowing for graceful shutdown, and a bugfix setting the default to `-1`:
* https://github.com/fluxcd/helm-controller/pull/570
* https://github.com/fluxcd/helm-controller/pull/582
